src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseStringMethodCharOverloadWithSingleCharacters.cs

```diff
@@ -0,0 +1,239 @@
+                    (comparison == ComparisonUsed.InvariantCulture && c.IsPrintableAscii()))
+                {
+                    rule = SafeTransformationRule;
+                }
+                // CA1866: Method(string)
+                else if (comparison == null)
+                {
+                    rule = NoSpecifiedComparisonRule;
+                }
+                // CA1867: Method(string, StringComparison.AnythingElse) or
+                //         Method(string, false/true, CultureInfo.AnythingElse)
+                else
+                {
+                    rule = AnyOtherSpecifiedComparisonRule;
+                }
+
+                var argumentList = GetArgumentList(stringArgument.Syntax);
+                if (argumentList != null)
+                {
+                    var arg0 = $"string.{method}(char)";
+                    var arg1 = $"string.{method}(string)";
+                    context.ReportDiagnostic(argumentList.CreateDiagnostic(rule, new[] { arg0, arg1 }));
+                }
+            }
+
+            bool TryMatchTargetMethod(
+                IInvocationOperation invocationOperation,
+                [NotNullWhen(true)] out string? method,
+                out ComparisonUsed? comparison)
+            {
+                method = null;
+                comparison = null;
+
+                if (invocationOperation.TargetMethod is IMethodSymbol invokedMethod &&
+                    invokedMethod.ContainingType.SpecialType == SpecialType.System_String &&
+                    TargetMethods.Contains(invokedMethod.Name) &&
+                    invokedMethod.Parameters.Length > 0 &&
+                    invokedMethod.Parameters[0].Type.SpecialType == SpecialType.System_String)
+                {
+                    method = invokedMethod.Name;
+
+                    foreach (var argument in invocationOperation.Arguments)
+                    {
+                        if (argument.Value.Type == null)
+                            continue;
+
+                        if (argument.Value.Type.Equals(stringComparisonType) &&
+                            argument.Value is IFieldReferenceOperation fieldReferenceOperation)
+                        {
+                            if (fieldReferenceOperation.Field.Equals(stringComparisonOrdinalSymbol))
+                            {
+                                comparison = ComparisonUsed.Ordinal;
+                            }
+                            else if (fieldReferenceOperation.Field.Equals(stringComparisonInvariantCultureSymbol))
+                            {
+                                comparison = ComparisonUsed.InvariantCulture;
+                            }
+                            else
+                            {
+                                comparison = ComparisonUsed.Other;
+                            }
+                        }
+
+                        if (argument.Value.Type.Equals(cultureInfoType) &&
```

Comment:
>if [](http://example.com/codeflow?start=24&length=2)

nit: would using `else` here be better to avoid the checks if you already have the signature that is using stringComparisonType?

Reply:
If you mean:
```cs
if (argument.Value.Type.Equals(stringComparisonType) &&
    argument.Value is IFieldReferenceOperation fieldReferenceOperation))
{
    ...
}
else // instead of if (argument.Value.Type.Equals(cultureInfoType) && argument.Value is IPropertyReferenceOperation propertyReferenceOperation)
{
    ...
}
```

Some of the target methods (such as `IndexOf`) have other kinds of parameters such as `int`, so can't have a bare `else` here.

Reply:
Yes, but this is inside a foreach over all the arguments (easier than trying to hard code indexes that also change with the overload):
```cs
foreach (var argument in invocationOperation.Arguments)
```

So this will go over the `string` arg, in addition to the `int`s in the case of `IndexOf`. Other than that, we would still need a `argument.Value is IPropertyReferenceOperation propertyReferenceOperation` check anyway.

Reply:
What I meant the else still will have the condition `if (argument.Value.Type.Equals(cultureInfoType) && argument.Value is IPropertyReferenceOperation propertyReferenceOperation)`. I didn't mean to remove this.

Reply:
Ah, you mean `else if`? Can do that of course. I thought you meant a bare `else`.